ui/raidboss/data/05-shb/eureka/delubrum_reginae_savage.js

```diff
@@ -596,6 +596,364 @@ export default {
+          de: 'Spaltteil der Eingeschworenen',
+          fr: 'Clone De La Trinité Féale',
+          ja: 'アヴァウドの分体',
+        };
+
+        // select the Trinity and Avatars
+        let combatantNameBoss = null;
+        let combatantNameAvatar = null;
+        combatantNameBoss = trinityLocaleNames[data.parserLang];
+        combatantNameAvatar = avatarLocaleNames[data.parserLang];
+
+        let combatantDataBoss = null;
+        let combatantDataAvatars = null;
+        if (combatantNameBoss) {
+          combatantDataBoss = await window.callOverlayHandler({
+            call: 'getCombatants',
+            names: [combatantNameBoss],
+          });
+        }
+        if (combatantNameAvatar) {
+          combatantDataAvatars = await window.callOverlayHandler({
+            call: 'getCombatants',
+            names: [combatantNameAvatar],
+          });
+        }
+
+        // if we could not retrieve combatant data, the
+        // trigger will not work, so just resume promise here
+        if (combatantDataBoss === null) {
+          console.error(`Trinity Avowed: null data`);
+          data.safeZone = null;
+          return;
+        }
+        if (!combatantDataBoss.combatants) {
+          console.error(`Trinity Avowed: null combatants`);
+          data.safeZone = null;
+          return;
+        }
+        if (combatantDataAvatars === null) {
+          console.error(`Avowed Avatar: null data`);
+          data.safeZone = null;
+          return;
+        }
+        if (!combatantDataAvatars.combatants) {
+          console.error(`Avowed Avatar: null combatants`);
+          data.safeZone = null;
+          return;
+        }
+        if (combatantDataAvatars.combatants.legnth < 3) {
+          console.error(`Avowed Avatar: expected at least 3 combatants got ${combatantDataAvatars.combatants.length}`);
+          data.safeZone = null;
+          return;
+        }
+
+        const getFacing = (combatant) => {
+          // Snap heading to closest card.
+          // N = 0, E = 1, S = 2, W = 3
+          return (2 - Math.round(combatant.Heading * 4 / Math.PI) / 2) % 4;
+        };
+
+        const getUnwaveringPosition = (combatant) => {
+          // Positions are moved downward 87 and left 277
+          const y = combatant.PosY + 87;
+          const x = combatant.PosX + 277;
+          // N = 0, E = 1, S = 2, W = 3
+          return Math.round(2 - 2 * Math.atan2(x, y) / Math.PI) % 4;
+        };
+
+        // we need to filter for the Trinity Avowed with the lowest ID
+        // that one is always cleaving on one of the cardinals
+        // Trinity Avowed is always East (-267, -87)
+        const eastCombatant =
+          combatantDataBoss.combatants.sort((a, b) => b.ID - a.ID).pop();
+
+        // we need to filter for the three Avowed Avatars with the lowest IDs
+        // as they cast cleave at the different cardinals
+        const avatarOne = combatantDataAvatars.combatants.sort((a, b) => b.ID - a.ID).pop();
```

Comment:
Do you need to sort these? How many avatars get these buffs?  Is it just the three?

Reply:
Just three avatars and the boss get the buff.

Unless the method to get the avatars is filtered differently, I think the sort is still required as logs indicate many more Avatars with different IDs in the encounter.

Reply:
May be possible to collect the info via a parseInt(matches.sourceId,16) during the buffs going out, adding enough delay there... but that gets to a more complicated race condition.

Reply:
It's a lot more complicated to code the collection off of the debuffs for matching to the adds than it is to sort, but I don't know the performance difference.

Reply:
Something like letting the trigger go off 4 times, each with its own promise and then a third trigger (or the last trigger to go off does a preRun to check if it is the one to execute the alert.

Reply:
I forgot to mention, but regarding what I said about using sourceId, you need to be careful when handling this if regex triggers off the boss, the boss (randomly?) gets its buff refreshed, so I had switched to having the trigger just be based on the Avatars getting it.

Reply:
Ah ok ok, thanks for all the details.  I think I am a little bit iffy on depending on actor ids when you don't have to, but I agree that it's probably more complicated to do other ways.  This is also kind of painful to test, so I think I'm inclined to just leave it as-is and get this landed.